delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/jsonld/JsonLdSourceParser.scala

```diff
@@ -18,51 +19,58 @@ import monix.bio.{IO, UIO}
+    * When no @id is present, one is generated using the base on the project suffixed with a randomly generated UUID.
+    *
+    * @param project the project with the base used to generate @id when needed and the @context when not provided on the source
     * @param source  the Json payload
+    * @return a tuple with the resulting @id iri and the decode value
     */
-  def computeId[R](project: Project, source: Json)(implicit
+  def decode[A: JsonLdDecoder, R](project: Project, source: Json)(implicit
       uuidF: UUIDF,
       rcr: RemoteContextResolution,
       rejectionMapper: Mapper[JsonLdRejection, R]
-  ): IO[R, Iri] =
+  ): IO[R, (Iri, A)] = {
     for {
-      (_, expanded) <- expandSource(project, source).leftMap(rejectionMapper.to)
+      (_, expanded) <- expandSource(project, source)
       iri           <- getOrGenerateId(expanded.rootId.asIri, project)
-    } yield iri
+      decodedValue  <- IO.fromEither(expanded.to[A].leftMap(DecodingFailed))
+    } yield (iri, decodedValue)
+  }.leftMap(rejectionMapper.to)
 
   /**
-    * Computes the id from the segment and validate it against the one in the source
+    * Expands the passed ''source'' and attempt to decode it into an ''A''
+    * The @id value is extracted from the payload if exists and compared to the passed ''iri''.
+    * If they aren't equal an [[UnexpectedId]] rejection is issued.
     *
-    * @param idSegment        the id provided in the segment
-    * @param project          the project
-    * @param source           the Json payload
+    * @param project the project with the base used to generate @id when needed and the @context when not provided on the source
+    * @param source  the Json payload
+    * @return a tuple with the resulting @id iri and the decode value
     */
-  def computeId[R](idSegment: IdSegment, project: Project, source: Json)(implicit
+  def decode[A: JsonLdDecoder, R](project: Project, iri: Iri, source: Json)(implicit
       rcr: RemoteContextResolution,
       rejectionMapper: Mapper[JsonLdRejection, R]
-  ): IO[R, Iri] =
+  ): IO[R, A] = {
     for {
-      (_, expanded) <- expandSource(project, source).leftMap(rejectionMapper.to)
-      iri           <- expandIri(idSegment, project)
-      _             <- checkAndSetSameId(iri, expanded).leftMap(rejectionMapper.to)
-    } yield iri
-
-  /**
-    * Expand the given segment to an Iri using the provided project if necessary
-    * @param segment the to translate to an Iri
-    * @param project the project
-    */
-  def expandIri[R](segment: IdSegment, project: Project)(implicit
-      rejectionMapper: Mapper[JsonLdRejection, R]
-  ): IO[R, Iri] =
-    IO.fromOption(
-      segment.toIri(project.apiMappings, project.base),
-      rejectionMapper.to(InvalidId(segment.asString))
-    )
+      (_, expanded) <- expandSource(project, source)
+      _             <- checkAndSetSameId(iri, expanded)
+      decodedValue  <- IO.fromEither(expanded.to[A].leftMap(DecodingFailed))
```

Comment:
you are using the `expanded` here, while you should be using the one resulting from the method `checkAndSetSameId `:
```scala
for {
   (_, originalExpanded) <- expandSource(project, source)
   expanded              <- checkAndSetSameId(iri, originalExpanded)
   decodedValue          <- IO.fromEither(expanded.to[A].leftMap(DecodingFailed))
} yield decodedValue
```